packages/gatsby/src/redux/__tests__/run-sift.js

```diff
@@ -62,136 +70,272 @@ if (!process.env.GATSBY_DB_NODES || process.env.GATSBY_DB_NODES === `redux`) {
+            queryArgs,
+            firstOnly: false,
+            nodeTypeNames: [gqlType.name],
+            typedKeyValueIndexes: createIndexCache(),
+          })
+
+          expect(resultSingular.map(o => o.id)).toEqual([mockNodes()[1].id])
+          expect(resultMany.map(o => o.id)).toEqual([mockNodes()[1].id])
         })
 
-        const resultMany = await runSift({
-          gqlType,
-          queryArgs,
-          firstOnly: false,
-          nodeTypeNames: [gqlType.name],
+        it(`eq operator honors type`, async () => {
+          const queryArgs = {
+            filter: {
+              id: { eq: `id_1` },
+            },
+          }
+
+          const resultSingular = await runSift({
+            gqlType,
+            queryArgs,
+            firstOnly: true,
+            nodeTypeNames: [gqlType.name],
+            typedKeyValueIndexes: createIndexCache(),
+          })
+
+          const resultMany = await runSift({
+            gqlType,
+            queryArgs,
+            firstOnly: false,
+            nodeTypeNames: [gqlType.name],
+            typedKeyValueIndexes: createIndexCache(),
+          })
+
+          // `id-1` node is not of queried type, so results should be empty
+          expect(resultSingular).toEqual([])
+          expect(resultMany).toEqual(null)
         })
 
-        expect(resultSingular.map(o => o.id)).toEqual([mockNodes()[1].id])
-        expect(resultMany.map(o => o.id)).toEqual([mockNodes()[1].id])
+        it(`non-eq operator`, async () => {
+          const queryArgs = {
+            filter: {
+              id: { ne: `id_2` },
+            },
+          }
+
+          const resultSingular = await runSift({
+            gqlType,
+            queryArgs,
+            firstOnly: true,
+            nodeTypeNames: [gqlType.name],
+            typedKeyValueIndexes: createIndexCache(),
+          })
+
+          const resultMany = await runSift({
+            gqlType,
+            queryArgs,
+            firstOnly: false,
+            nodeTypeNames: [gqlType.name],
+            typedKeyValueIndexes: createIndexCache(),
+          })
+
+          expect(resultSingular.map(o => o.id)).toEqual([mockNodes()[2].id])
+          expect(resultMany.map(o => o.id)).toEqual([
+            mockNodes()[2].id,
+            mockNodes()[3].id,
+          ])
+        })
+        it(`return empty array in case of empty nodes`, async () => {
+          const queryArgs = { filter: {}, sort: {} }
+          const resultSingular = await runSift({
+            gqlType,
+            queryArgs,
+            firstOnly: true,
+            nodeTypeNames: [`NonExistentNodeType`],
+            typedKeyValueIndexes: createIndexCache(),
+          })
+          expect(resultSingular).toEqual([])
+        })
       })
+      describe(`filters by arbitrary property correctly`, () => {
+        it(`eq operator flat single`, async () => {
+          const queryArgs = {
+            filter: {
+              slog: { eq: `def` },
```

Comment:
[nit] I know it is only test code. 

But did you mean `slug` here?

Reply:
To be honest, no. This approach is generic, so not for `id` and not for the very common name `slug`, so I went with `slog` instead. Maybe I should have just gone with `foo` instead... feel free to PR that if it bothers you :)